docs/testing-guidelines.md

```diff
@@ -0,0 +1,121 @@
+- Handle limitations due to licensing or other constraints by documenting them and implementing what testing is possible.
+
+### **Tools**
+- UI Testing: Playwright, Cypress
+- Non-UI Testing: Jest
+- Custom Scripts: Bash or other scripting languages as needed
+
+## Upgrade Testing
+Ensure that the current development package works when deployed over the previously released version.
+
+### **Key Considerations**
+- Test data migration and persistence.
+- Verify that configurations are properly carried over or updated.
+- Check for any breaking changes in APIs or integrations.
+
+## Linting and Static Analysis
+
+Use appropriate tooling to ensure code quality and consistency.
+
+### **Recommended Tools**
+- [YAML Linting](https://github.com/adrienverge/yamllint): `yamllint`
+- [Zarf Package Linting](https://docs.zarf.dev/commands/zarf_dev_lint/): `zarf dev lint`
+- [Shellcheck](https://www.shellcheck.net/) for any bash scripts.
+
+## **Best Practices**
+
+1. **Consistency**: Use similar testing approaches across different UDS packages where possible.
+2. **Documentation**: Clearly document what each test is checking and why.
+3. **Maintenance**: Keep tests updated as the application and deployment methods evolve.
+4. **CI/CD Integration**: Incorporate all types of tests into your CI/CD pipeline.
+5. **Error Handling**: Ensure tests fail gracefully and provide clear error messages.
+6. **Repeatability**: Ensure tests produce consistent results regardless of how many times or in what order they are executed. Design them to handle dynamic and asynchronous workloads without compromising output integrity.
+7. **Performance Consideration**: Balance thoroughness with rapid developer feedback:
+    - **Single package tests:** Aim for <15 minutes runtime
+    - **Full CI jobs** (e.g., deploy and test): Maximum 30 minutes
+    - **Large integration tests** (multiple packages): Longer durations acceptable
+
+    > **Note:** Strive for individual package tests to complete in under 10 minutes when possible.
+
+## **Best Practice Examples**
+The following examples demonstrate best practices for implementing various types of tests in UDS packages:
+
+- ### **Journey Tests**
+    ---
+    The `uds-package-gitlab` repository provides a good example of journey test implementation.
+    These files demonstrate how to set up and structure end-to-end tests using Playwright, including authentication handling and test suite organization.
+
+    - **Test Configuration**: [playwright.config.ts](https://github.com/defenseunicorns/uds-package-gitlab/blob/main/tests/playwright.config.ts)
+    - **Test Suite**: [gitlab.test.ts](https://github.com/defenseunicorns/uds-package-gitlab/blob/main/tests/gitlab.test.ts)
+    - **Authentication Setup**: [auth.setup.ts](https://github.com/defenseunicorns/uds-package-gitlab/blob/main/tests/auth.setup.ts)
+
+- ### Upgrade Tests
+    ---
+    (TODO: Add specific examples of upgrade tests from UDS packages)
```

Comment:
Yes!

This UDS common action is reused in packages: 
https://github.com/defenseunicorns/uds-common/blob/main/.github/actions/test-deploy/action.yaml

Example (everyone's favorite Gitlab):
https://github.com/defenseunicorns/uds-package-gitlab/blob/main/tasks.yaml#L48-L68

Reply:
this was done here: https://github.com/defenseunicorns/uds-common/pull/205/commits/355f2ca6b82408a837655fc19f367cc90b56cc7b